rust/protocol/src/session_cipher.rs

```diff
@@ -296,21 +312,53 @@ fn decrypt_message_with_record<R: Rng + CryptoRng>(
     ciphertext: &SignalMessage,
     csprng: &mut R,
 ) -> Result<Vec<u8>> {
+    let log_decryption_failure = |state: &SessionState, error: &SignalProtocolError| {
+        log::error!(
+            "Failed to decrypt whisper message with ratchet key: {} and counter: {}. \
+             Session loaded for {}. Local session has base key: {} and counter: {}. {}",
+            hex::encode(ciphertext.sender_ratchet_key().public_key_bytes().unwrap()),
+            ciphertext.counter(),
+            remote_address,
+            hex::encode(
+                state
+                    .sender_ratchet_key()
+                    .ok()
+                    .as_ref()
+                    .and_then(|k| k.public_key_bytes().ok())
+                    .unwrap_or(&[])
+            ),
+            state.previous_counter().unwrap_or(u32::MAX),
+            error
+        );
+    };
+
     let mut errs = vec![];
 
     if let Ok(current_state) = record.session_state() {
         let mut current_state = current_state.clone();
-        let result = decrypt_message_with_state(&mut current_state, ciphertext, csprng);
+        let result =
+            decrypt_message_with_state(&mut current_state, ciphertext, remote_address, csprng);
 
         match result {
             Ok(ptext) => {
+                log::debug!(
+                    "successfully decrypted with current session state (base key {})",
+                    hex::encode(
+                        current_state
+                            .sender_ratchet_key()
+                            .unwrap()
+                            .public_key_bytes()
+                            .unwrap()
```

Comment:
Again concerned about the unwraps though these are probably safe since if we were able to decrypt the key is probably there...

Reply:
This specific value is logged regularly - maybe add a `fn` to `SessionState` to handle it?